[Hxb] Optimistic display requests

kLabz opened the pull request:
Limits the number of modules restored during display requests, which is especially useful on projects with lots of inter dependencies.

* Mostly reverts #11650: instead of looking at what hxb writer saw when adding a module to the cache, looks at what hxb reader actually used
* Built on top of #11660 (and #11872 that fixes some of its issues)
* ~~Similar to #11873, uses `TLazy` on `cf_type`. Initially, modules that are non essential are only partially loaded. `cf_type` is being followed when a field is used, which will trigger the loading of the rest of the module.~~
* Can be disabled with `-D disable-hxb-optimizations` in case that causes issues

kLabz committed:
[tests] update test

kLabz committed:
[hxb] load less dependencies in display requests

Can be disabled with -D disable-hxb-optimizations

kLabz committed:
[CI] run server tests with different cache configurations

Simn commented:
Oh dear...

kLabz commented:
I will rebase after #11872 and #11873 are merged to get a cleaner PR here.

Will keep it opt-in (and a draft) for now, still waiting on more data from shiro devs and will likely have a thing or two to fix.

kLabz committed:
[hxb] only load up to EOF if at least one field is accessed (or full typing)

kLabz committed:
[hxb] read/write field metadata earlier

Simn commented on src-json/define.json:
```diff
@@ -71,6 +71,11 @@
+	{
+		"name": "DisableHxbOptimizations",
+		"define": "disable-hxb-optimizations",
+		"doc": "Disable shortcuts hxb cache uses to speed up display requests."
```
Somehow I had an aneurysm while reading this, though now that I know what it means it makes sense. Still, "Disable shortcuts used by hxb cache to speed up display requests." might make it easier to parse.

kLabz committed:
[hxb] disable hxb optimizations when hxb cache is disabled..

kLabz committed:
[hxb] delay to PForce pass

kLabz committed:
Reduce diff and cleanup a bit

kLabz committed:
Revert "[hxb] only load up to EOF if at least one field is accessed (or full typing)"

This reverts commit 091cd94c3f95375ad89545f0b61f9c2c566bef82.

kLabz committed:
Only delay make_lazy_type with PForce; use PConnectField for delaying chunks reading

kLabz committed:
[hxb] disable lazy wrapping for abstracts for now

Something's wrong there, getting Dynamic fields in display requests.
Need to investigate and add that lazy wrapping again in another PR.

kLabz committed:
Extract typer_pass to avoid extra dependencies
